File: sdk-source/java/src/main/java/io/aegis/sdk/Json.java
package io.aegis.sdk;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** Minimal dependency-free JSON reader/writer used by the SDK. */
final class Json {
    private final String source;
    private int index;

    private Json(String source) {
        this.source = source;
    }

    static Object parse(String text) {
        Json parser = new Json(text == null || text.isBlank() ? "{}" : text);
        parser.skipWhitespace();
        Object value = parser.readValue();
        return value;
    }

    @SuppressWarnings("unchecked")
    static Map<String, Object> parseObject(String text) {
        Object value = parse(text);
        return value instanceof Map ? (Map<String, Object>) value : new LinkedHashMap<>();
    }

    static String write(Object value) {
        StringBuilder out = new StringBuilder();
        writeValue(value, out);
        return out.toString();
    }

    private static void writeValue(Object value, StringBuilder out) {
        if (value == null) {
            out.append("null");
        } else if (value instanceof String string) {
            writeString(string, out);
        } else if (value instanceof Number || value instanceof Boolean) {
            out.append(value);
        } else if (value instanceof Map<?, ?> map) {
            out.append('{');
            boolean first = true;
            for (Map.Entry<?, ?> entry : map.entrySet()) {
                if (entry.getValue() == null) continue;
                if (!first) out.append(',');
                first = false;
                writeString(String.valueOf(entry.getKey()), out);
                out.append(':');
                writeValue(entry.getValue(), out);
            }
            out.append('}');
        } else if (value instanceof Iterable<?> items) {
            out.append('[');
            boolean first = true;
            for (Object item : items) {
                if (!first) out.append(',');
                first = false;
                writeValue(item, out);
            }
            out.append(']');
        } else {
            writeString(String.valueOf(value), out);
        }
    }

    private static void writeString(String value, StringBuilder out) {
        out.append('"');
        for (int i = 0; i < value.length(); i++) {
            char c = value.charAt(i);
            switch (c) {
                case '"' -> out.append("\\\"");
                case '\\' -> out.append("\\\\");
                case '\n' -> out.append("\\n");
                case '\r' -> out.append("\\r");
                case '\t' -> out.append("\\t");
                default -> {
                    if (c < 0x20) {
                        out.append(String.format("\\u%04x", (int) c));
                    } else {
                        out.append(c);
                    }
                }
            }
        }
        out.append('"');
    }

    private Object readValue() {
        skipWhitespace();
        if (index >= source.length()) return null;
        char c = source.charAt(index);
        return switch (c) {
            case '{' -> readObject();
            case '[' -> readArray();
            case '"' -> readString();
            case 't' -> readLiteral("true", Boolean.TRUE);
            case 'f' -> readLiteral("false", Boolean.FALSE);
            case 'n' -> readLiteral("null", null);
            default -> readNumber();
        };
    }

    private Map<String, Object> readObject() {
        Map<String, Object> map = new LinkedHashMap<>();
        index++; // {
        skipWhitespace();
        if (peek() == '}') { index++; return map; }
        while (index < source.length()) {
            skipWhitespace();
            String key = readString();
            skipWhitespace();
            index++; // :
            map.put(key, readValue());
            skipWhitespace();
            char c = peek();
            index++;
            if (c == '}') break;
        }
        return map;
    }

    private List<Object> readArray() {
        List<Object> list = new ArrayList<>();
        index++; // [
        skipWhitespace();
        if (peek() == ']') { index++; return list; }
        while (index < source.length()) {
            list.add(readValue());
            skipWhitespace();
            char c = peek();
            index++;
            if (c == ']') break;
        }
        return list;
    }

    private String readString() {
        StringBuilder out = new StringBuilder();
        index++; // opening quote
        while (index < source.length()) {
            char c = source.charAt(index++);
            if (c == '"') break;
            if (c == '\\') {
                char escaped = source.charAt(index++);
                switch (escaped) {
                    case 'n' -> out.append('\n');
                    case 'r' -> out.append('\r');
                    case 't' -> out.append('\t');
                    case 'b' -> out.append('\b');
                    case 'f' -> out.append('\f');
                    case 'u' -> {
                        out.append((char) Integer.parseInt(source.substring(index, index + 4), 16));
                        index += 4;
                    }
                    default -> out.append(escaped);
                }
            } else {
                out.append(c);
            }
        }
        return out.toString();
    }

    private Object readLiteral(String literal, Object value) {
        index += literal.length();
        return value;
    }

    private Object readNumber() {
        int start = index;
        while (index < source.length() && "-+.eE0123456789".indexOf(source.charAt(index)) >= 0) index++;
        String raw = source.substring(start, index);
        if (raw.contains(".") || raw.contains("e") || raw.contains("E")) return Double.parseDouble(raw);
        return Long.parseLong(raw);
    }

    private char peek() {
        return index < source.length() ? source.charAt(index) : '\0';
    }

    private void skipWhitespace() {
        while (index < source.length() && Character.isWhitespace(source.charAt(index))) index++;
    }
}

File: sdk-source/kotlin/src/main/kotlin/io/aegis/sdk/Json.kt
package io.aegis.sdk

/**
 * Minimal JSON encoder/decoder so the SDK stays dependency-free.
 * Decodes into Map/List/String/Double/Boolean/null.
 */
internal object Json {

    fun encode(value: Any?): String = when (value) {
        null -> "null"
        is String -> encodeString(value)
        is Boolean -> value.toString()
        is Number -> value.toString()
        is Map<*, *> -> value.entries.filter { it.value != null }
            .joinToString(",", "{", "}") { "${encodeString(it.key.toString())}:${encode(it.value)}" }
        is Iterable<*> -> value.joinToString(",", "[", "]") { encode(it) }
        else -> encodeString(value.toString())
    }

    fun decode(text: String): Any? {
        if (text.isBlank()) return emptyMap<String, Any?>()
        val cursor = Cursor(text)
        return cursor.readValue()
    }

    private fun encodeString(value: String): String {
        val sb = StringBuilder("\"")
        for (c in value) {
            when (c) {
                '"' -> sb.append("\\\"")
                '\\' -> sb.append("\\\\")
                '\n' -> sb.append("\\n")
                '\r' -> sb.append("\\r")
                '\t' -> sb.append("\\t")
                else -> if (c < ' ') sb.append("\\u%04x".format(c.code)) else sb.append(c)
            }
        }
        return sb.append('"').toString()
    }

    private class Cursor(val text: String) {
        var i = 0

        fun readValue(): Any? {
            skip()
            return when {
                i >= text.length -> null
                text[i] == '{' -> readObject()
                text[i] == '[' -> readArray()
                text[i] == '"' -> readString()
                text.startsWith("true", i) -> { i += 4; true }
                text.startsWith("false", i) -> { i += 5; false }
                text.startsWith("null", i) -> { i += 4; null }
                else -> readNumber()
            }
        }

        fun skip() {
            while (i < text.length && text[i].isWhitespace()) i++
        }

        fun readObject(): Map<String, Any?> {
            val out = LinkedHashMap<String, Any?>()
            i++
            skip()
            if (i < text.length && text[i] == '}') { i++; return out }
            while (i < text.length) {
                skip()
                val key = readString()
                skip()
                if (i < text.length && text[i] == ':') i++
                out[key] = readValue()
                skip()
                when {
                    i < text.length && text[i] == ',' -> i++
                    i < text.length && text[i] == '}' -> { i++; break }
                    else -> break
                }
            }
            return out
        }

        fun readArray(): List<Any?> {
            val out = ArrayList<Any?>()
            i++
            skip()
            if (i < text.length && text[i] == ']') { i++; return out }
            while (i < text.length) {
                out.add(readValue())
                skip()
                when {
                    i < text.length && text[i] == ',' -> i++
                    i < text.length && text[i] == ']' -> { i++; break }
                    else -> break
                }
            }
            return out
        }

        fun readString(): String {
            if (text[i] != '"') return ""
            i++
            val sb = StringBuilder()
            while (i < text.length && text[i] != '"') {
                if (text[i] == '\\' && i + 1 < text.length) {
                    i++
                    when (val esc = text[i]) {
                        'n' -> sb.append('\n')
                        't' -> sb.append('\t')
                        'r' -> sb.append('\r')
                        'b' -> sb.append('\b')
                        'u' -> { sb.append(text.substring(i + 1, i + 5).toInt(16).toChar()); i += 4 }
                        else -> sb.append(esc)
                    }
                } else {
                    sb.append(text[i])
                }
                i++
            }
            i++
            return sb.toString()
        }

        fun readNumber(): Double {
            val start = i
            while (i < text.length && (text[i].isDigit() || text[i] in "-+.eE")) i++
            return text.substring(start, i).toDoubleOrNull() ?: 0.0
        }
    }
}

/** Convenience accessors for decoded JSON maps. */
@Suppress("UNCHECKED_CAST")
fun Map<String, Any?>.obj(key: String): Map<String, Any?> = this[key] as? Map<String, Any?> ?: emptyMap()
fun Map<String, Any?>.str(key: String, fallback: String = ""): String = this[key] as? String ?: fallback
fun Map<String, Any?>.bool(key: String, fallback: Boolean = false): Boolean = this[key] as? Boolean ?: fallback